akka-actor/src/main/scala/akka/util/ByteString.scala

```diff
@@ -253,9 +253,11 @@ object ByteString {
       else ByteString1(bytes, startIndex, Math.min(n, length))
 
     override def slice(from: Int, until: Int): ByteString = {
-      if (from <= 0 && until >= length) this // we can do < / > since we're Compact
+      if (from <= 0 && length <= until) this // we can do < / > since we're Compact
       else if (until <= from) ByteString1.empty
-      else ByteString1(bytes, startIndex + from, until - from)
+      else if (0 <= from && until <= bytes.length) ByteString1(bytes, startIndex + from, until - from)
+      else if (from < 0) ByteString1(bytes, startIndex, until)
+      else /*(until > bytes.length)*/ ByteString1(bytes, startIndex + from, bytes.length - from)
     }
```

Comment:
Also crashes on some scalachecks.

I went ahead and checked if we actually can just drop it and do it via `take/drop` turns out perf is good enough:

was (but broken):

```
   [info] ByteString_dropSliceTake_Benchmark.bs_large_slice_80_80          thrpt   20  233 017 314.148 ± 7070246.826  ops/s
   [info] ByteString_dropSliceTake_Benchmark.bs_large_slice_129_129        thrpt   20  275 245 086.247 ± 4969752.048  ops/s
   [info] ByteString_dropSliceTake_Benchmark.bss_large_slice_80_80         thrpt   20  264 963 420.976 ± 4259289.143  ops/s
   [info] ByteString_dropSliceTake_Benchmark.bss_large_slice_129_129       thrpt   20  265 477 577.022 ± 4623974.283  ops/s
```

is (and passes everything):

```
[info] ByteString_dropSliceTake_Benchmark.bs_large_slice_129_129   thrpt   40  213658205.432 ± 5877072.789  ops/s
[info] ByteString_dropSliceTake_Benchmark.bs_large_slice_80_80     thrpt   40  209119190.836 ± 6837219.988  ops/s
[info] ByteString_dropSliceTake_Benchmark.bss_large_slice_129_129  thrpt   40  265664253.164 ± 4064734.763  ops/s
[info] ByteString_dropSliceTake_Benchmark.bss_large_slice_80_80    thrpt   40  260233630.884 ± 5312998.911  ops/s
```

I'm tempted to call that good enough for now as we want to cut a fix-release ASAP actually :~
Unless you have more time on your hands to keep figuring out the optimal way here @monkey-mas.

I also added the following edge case tests so that they always run:

```
      ByteStrings(ByteString1.fromString("a"), ByteString1.fromString("")).slice(1, -2) should ===(ByteString(""))
      ByteStrings(ByteString1.fromString("a"), ByteString1.fromString("")).slice(-10, -100) should ===(ByteString(""))
      ByteStrings(ByteString1.fromString("a"), ByteString1.fromString("")).slice(-100, -10) should ===(ByteString(""))
      ByteString1.fromString("cd").slice(100, 10) should ===(ByteString(""))
      ByteString1.fromString("cd").slice(100, 1000) should ===(ByteString(""))
      ByteString1.fromString("cd").slice(-10, -5) should ===(ByteString(""))
      ByteString1.fromString("cd").slice(-2, -5) should ===(ByteString(""))
      ByteString1.fromString("cd").slice(-2, 1) should ===(ByteString("c"))
```


if you want to hack on it more I suggest you use the following config btw:

```scala
  implicit val betterGeneratorDrivenConfig = PropertyCheckConfig().copy(minSuccessful = 1000) 
// just add it on top of the file
```

it found much more corner cases for me than the default settings.